src/decode-vxlan.c

```diff
@@ -153,37 +165,60 @@ int DecodeVXLAN(ThreadVars *tv, DecodeThreadVars *dtv, Packet *p,
-        case ETHERNET_TYPE_8021AD:
-        case ETHERNET_TYPE_8021QINQ:
-            SCLogDebug("VXLAN found VLAN");
-            decode_tunnel_proto = DECODE_TUNNEL_VLAN;
-            break;
-        default:
-            SCLogDebug("VXLAN found unsupported Ethertype - expected IPv4, IPv6, VLAN, or ARP");
-            ENGINE_SET_INVALID_EVENT(p, VXLAN_UNKNOWN_PAYLOAD_TYPE);
+    /* TODO check further that we strip only once per real packet */
+    enum TunnelPacketMode m = TUNNEL_PACKET_MODE_SPLIT;
+    if ((p->flags & PKT_TUNNEL_STRIPPED) == 0 &&
+            g_vxlan_tunnel_pkt_mode == TUNNEL_PACKET_MODE_STRIP_ENV) {
+        m = TUNNEL_PACKET_MODE_STRIP_ENV;
     }
 
-    /* Set-up and process inner packet if it is a supported ethertype */
-    if (decode_tunnel_proto != DECODE_TUNNEL_UNSET) {
-        Packet *tp = PacketTunnelPktSetup(tv, dtv, p, pkt + VXLAN_HEADER_LEN + ETHERNET_HEADER_LEN,
-                len - (VXLAN_HEADER_LEN + ETHERNET_HEADER_LEN), decode_tunnel_proto);
-        if (tp != NULL) {
-            PKT_SET_SRC(tp, PKT_SRC_DECODER_VXLAN);
-            PacketEnqueueNoLock(&tv->decode_pq, tp);
+    if (m == TUNNEL_PACKET_MODE_SPLIT) {
+        switch (eth_type) {
+            case ETHERNET_TYPE_ARP:
+                SCLogDebug("VXLAN found ARP");
+                break;
+            case ETHERNET_TYPE_IP:
+                SCLogDebug("VXLAN found IPv4");
+                decode_tunnel_proto = DECODE_TUNNEL_IPV4;
+                break;
+            case ETHERNET_TYPE_IPV6:
+                SCLogDebug("VXLAN found IPv6");
+                decode_tunnel_proto = DECODE_TUNNEL_IPV6;
+                break;
+            case ETHERNET_TYPE_VLAN:
+            case ETHERNET_TYPE_8021AD:
+            case ETHERNET_TYPE_8021QINQ:
+                SCLogDebug("VXLAN found VLAN");
+                decode_tunnel_proto = DECODE_TUNNEL_VLAN;
+                break;
+            default:
+                SCLogDebug("VXLAN found unsupported Ethertype - expected IPv4, IPv6, VLAN, or ARP");
+                ENGINE_SET_INVALID_EVENT(p, VXLAN_UNKNOWN_PAYLOAD_TYPE);
+        }
+
+        /* Set-up and process inner packet if it is a supported ethertype */
+        if (decode_tunnel_proto != DECODE_TUNNEL_UNSET) {
+            Packet *tp =
+                    PacketTunnelPktSetup(tv, dtv, p, pkt + VXLAN_HEADER_LEN + ETHERNET_HEADER_LEN,
+                            len - (VXLAN_HEADER_LEN + ETHERNET_HEADER_LEN), decode_tunnel_proto);
+            if (tp != NULL) {
+                PKT_SET_SRC(tp, PKT_SRC_DECODER_VXLAN);
+                PacketEnqueueNoLock(&tv->decode_pq, tp);
+            }
         }
+    } else if (m == TUNNEL_PACKET_MODE_STRIP_ENV) {
+        uint32_t encap_size = len - VXLAN_HEADER_LEN;
+        const uint8_t *encap_pkt = pkt + VXLAN_HEADER_LEN;
+
+        /* reset existing layers */
+        PacketClearL2(p);
+        PacketClearL3(p);
+        PacketClearL4(p);
+        // TODO clear things more: tuple, events?
```

Comment:
need to clear more of the packet: addr, port, ipproto, decode events. Need to review flags too.